// eslint-config-algolia/rules/react.js
// eslint-disable-next-line import/no-commonjs
module.exports = {
  rules: {
    'react/no-danger': ['off'],
    'react/display-name': ['off'],
    'react/jsx-key': ['error'],
    'react/jsx-no-bind': [
      'error',
      {
        allowArrowFunctions: true,
      },
    ],
    'react-hooks/rules-of-hooks': ['error'],
    'react-hooks/exhaustive-deps': ['warn'],

    'react/button-has-type': ['error'],
    'react/destructuring-assignment': ['off'],
    'react/jsx-boolean-value': ['error', 'always'],
    'react/jsx-filename-extension': ['error', { extensions: ['.tsx', '.jsx'] }],
    'react/jsx-pascal-case': ['error'],
    'react/jsx-no-target-blank': ['error'],
    'react/jsx-wrap-multilines': [
      'error',
      {
        declaration: 'parens',
        assignment: 'parens',
        return: 'parens',
        arrow: 'parens',
        condition: 'ignore',
        logical: 'ignore',
        prop: 'ignore',
      },
    ],
    'react/jsx-props-no-spreading': ['off'],
    'react/jsx-fragments': ['error'],
    'react/jsx-handler-names': ['error'],
    'react/no-access-state-in-setstate': ['error'],
    'react/no-unescaped-entities': ['error'],
    'react/no-unused-state': ['error'],
    'react/prop-types': ['off'],
    'react/sort-comp': ['off'],
    'react/static-property-placement': ['error'],
    'react/function-component-definition': [
      'error',
      {
        namedComponents: 'arrow-function',
      },
    ],
    'react/no-array-index-key': ['warn'],
    'react/no-deprecated': ['error'],
    'react/no-did-mount-set-state': ['warn'],
    'react/no-did-update-set-state': ['warn'],
    'react/no-direct-mutation-state': ['error'],
    'react/no-find-dom-node': ['error'],
    'react/no-is-mounted': ['error'],
    'react/no-multi-comp': ['warn', { ignoreStateless: true }],
    'react/no-redundant-should-component-update': ['error'],
    'react/no-string-refs': ['error'],
    'react/no-this-in-sfc': ['error'],
    'react/no-typos': ['error'],
    'react/prefer-stateless-function': ['warn'],
    'react/self-closing-comp': ['error', { html: false }],
    'react/jsx-sort-props': [
      'error',
      {
        callbacksLast: true,
        shorthandFirst: true,
        shorthandLast: false,
        ignoreCase: true,
        noSortAlphabetically: true,
        reservedFirst: false,
      },
    ],
    'react/jsx-uses-react': 'error',
    'react/jsx-uses-vars': 'error',
    'react/react-in-jsx-scope': 'error',

    // ---- JSX-a11y
    'jsx-a11y/alt-text': ['error'],
    'jsx-a11y/anchor-is-valid': ['error'],
    'jsx-a11y/anchor-has-content': ['error', { components: [] }],
    'jsx-a11y/aria-activedescendant-has-tabindex': ['error'],
    'jsx-a11y/aria-props': ['error'],
    'jsx-a11y/aria-proptypes': ['error'],
    'jsx-a11y/aria-role': ['error', { ignoreNonDom: false }],
    'jsx-a11y/click-events-have-key-events': ['off'],
    'jsx-a11y/control-has-associated-label': ['off'],
    'jsx-a11y/heading-has-content': ['off'],
    'jsx-a11y/html-has-lang': ['error'],
    'jsx-a11y/iframe-has-title': ['error'],
    'jsx-a11y/interactive-supports-focus': ['error'],
    'jsx-a11y/label-has-associated-control': ['off'],
    'jsx-a11y/lang': ['error'],
    'jsx-a11y/media-has-caption': ['error'],
    'jsx-a11y/mouse-events-have-key-events': ['error'],
    'jsx-a11y/no-access-key': ['error'],
    'jsx-a11y/no-autofocus': ['error', { ignoreNonDOM: true }],
    'jsx-a11y/no-distracting-elements': [
      'error',
      {
        elements: ['marquee', 'blink'],
      },
    ],
    'jsx-a11y/no-interactive-element-to-noninteractive-role': ['error'],
    'jsx-a11y/no-noninteractive-element-interactions': ['error'],
    'jsx-a11y/no-noninteractive-element-to-interactive-role': ['error'],
    'jsx-a11y/no-noninteractive-tabindex': ['error'],
    'jsx-a11y/no-onchange': ['error'],
    'jsx-a11y/no-redundant-roles': ['error'],
    'jsx-a11y/no-static-element-interactions': ['error'],
    'jsx-a11y/role-has-required-aria-props': ['error'],
    'jsx-a11y/role-supports-aria-props': ['error'],
    'jsx-a11y/scope': ['error'],
    'jsx-a11y/tabindex-no-positive': ['error'],
  },
};
